fedimint/src/setup/mod.rs

```diff
@@ -0,0 +1,231 @@
+
+    // run fedimint
+    run_fedimint(&mut state);
+
+    Ok(Redirect::to("/".parse().unwrap()))
+}
+
+fn save_configs(server_config: &ServerConfig, client_config: &ClientConfig, cfg_path: &PathBuf) {
+    // Recursively create config directory if it doesn't exist
+    let parent = cfg_path.parent().unwrap();
+    std::fs::create_dir_all(&parent).expect("Failed to create config directory");
+
+    // Save the configs
+    tracing::info!("{:?}", cfg_path);
+    let cfg_file = std::fs::File::create(&cfg_path).expect("Could not create cfg file");
+    serde_json::to_writer_pretty(cfg_file, &server_config).unwrap();
+    let client_cfg_path = parent.join("client.json");
+    let client_cfg_file =
+        std::fs::File::create(&client_cfg_path).expect("Could not create cfg file");
+    serde_json::to_writer_pretty(client_cfg_file, &client_config).unwrap();
+}
+
+#[derive(Template)]
+#[template(path = "configs.html")]
+struct DisplayConfigsTemplate {
+    server_configs: Vec<(Guardian, String)>,
+    client_config: String,
+}
+
+async fn display_configs(Extension(state): Extension<MutableState>) -> DisplayConfigsTemplate {
+    let state = state.read().unwrap();
+    let server_configs = state
+        .server_configs
+        .clone()
+        .unwrap()
+        .into_iter()
+        .map(|(guardian, cfg)| (guardian, serde_json::to_string(&cfg).unwrap()))
+        .collect();
+    DisplayConfigsTemplate {
+        server_configs,
+        client_config: serde_json::to_string(&state.client_config.as_ref().unwrap()).unwrap(),
+    }
+}
+
+async fn qr(Extension(state): Extension<MutableState>) -> impl axum::response::IntoResponse {
+    let client_config = state.read().unwrap().client_config.clone().unwrap();
+    let connect_info = WsFederationConnect::from(&client_config);
+    let string = serde_json::to_string(&connect_info).unwrap();
+    // this was a hack to do a remote demo ... leaving just in case I need to do another!
+    // .replace("127.0.0.1", "188.166.55.8");
+    let png_bytes: Vec<u8> = qrcode_generator::to_png_to_vec(string, QrCodeEcc::Low, 1024).unwrap();
+    (
+        axum::response::Headers([(axum::http::header::CONTENT_TYPE, "image/png")]),
+        png_bytes,
+    )
+}
+
+// TODO: write cfg_path and db_path into state so we don't re-compute them
+#[derive(Debug)]
+struct State {
+    guardians: Vec<Guardian>,
+    // TODO: map name to peer id
+    running: bool,
+    cfg_path: PathBuf,
+    connection_string: String,
+    sender: Sender<()>,
+    server_configs: Option<Vec<(Guardian, ServerConfig)>>,
+    client_config: Option<ClientConfig>,
+}
+type MutableState = Arc<RwLock<State>>;
+
+pub async fn run_ui_setup(cfg_path: PathBuf, port: u16, sender: Sender<()>) {
+    let mut rng = OsRng::new().unwrap();
+    let secp = bitcoin::secp256k1::Secp256k1::new();
+    let (_, pubkey) = secp.generate_keypair(&mut rng);
+    let our_ip = "127.0.0.1"; // TODO: get our actual IP ... or pass in as argument???
```

Comment:
We should be able to get our public IP here. If we can give our public ip to other servers, then this should work across different machines. @elsirion what is the best way to get our public IP? Is something like https://docs.rs/public-ip/latest/public_ip completely unacceptable??

Reply:
Another command line argument to fedimintd???

Reply:
Spoke with @elsirion. Let's just ask the user for the IP in the UI. This is less error prone than automated approach. Can refine later. Could do this as follow-on PR, too.